### DISCUSSION TITLE
Is there any way to remove drawings?

### DISCUSSION THREAD

**Discussion by huyangsh** [pymupdf/PyMuPDF#865] [Looking for help] [answered]
**Is your feature request related to a problem? Please describe.**
I would like to remove some watermark-like contents from a pdf file. These contents are hard-drawn as drawings (paths). I have read the document, but I have not seen any methods like "deleteDrawing".

**Describe the solution you'd like**
If possible, maybe a new interface for deleting drawings could be provided.

**Describe alternatives you've considered**
Maybe redact annotations can be extended to support deletion of drawings, as they work fine for deleting texts. I do not see a specific reason not to add this functionality.

**Comment by JorjMcKie** [comment]
As you know, PyMuPDF is a binding to the library MuPDF, and as such is limited in its features to what MuPDF exposes as base features. Among such features are redaction annotations, and they simply don't cover removing watermarks.

Maybe you should submit a request to MuPDF: https://bugs.ghostscript.com/enter_bug.cgi.

On the other hand, you sound so confident:

> I do not see a specific reason not to add this functionality.

Do you have a specific idea in mind?

**Comment by JorjMcKie** [comment]
If you lookup the definition of redact annots, you will find, that the are supposed to remove **_everything_** inside a given retangle.

So no matter what they are capable of: they won't **_ever_** do what you want ... if they could remove a watermark, they would also remove overlapping text, images, etc.

**Comment by JorjMcKie** [comment]
If you can share your PDF (**please** no giant file ... maybe extract 1 page only), I am willing to have a look at it.

**Comment by user202729** [comment, marked as answer]
You kind of have to learn the PDF specification for these tasks (warning: it's huge...)

Then, the idea would be something like (extract the content object of the page, parse the data, determine if a drawing command draws on the watermark, remove them)

Or load the PDF up in some (commercial?) GUI PDF debugger to figure out which object it is.
